drivers/soundwire/generic_bandwidth_allocation.c

```diff
@@ -370,16 +402,91 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
 	}
 
 	if (i == clk_values) {
-		dev_err(bus->dev, "%s: could not find clock value for bandwidth %d\n",
+multilane:
+		dev_dbg(bus->dev, "%s: could not find clock value for bandwidth %d, checking multi-lane\n",
 			__func__, bus->params.bandwidth);
-		return -EINVAL;
+
+		m_rt = list_last_entry(&bus->m_rt_list, struct sdw_master_runtime, bus_node);
+		/*
+		 * Find available manager lanes that connected to the first Peripheral. No need
+		 * to check all Peripherals available lanes because we can't use multi-lane
+		 * if we can't find any available lane for the first Peripheral.
+		 */
+		s_rt = list_first_entry(&m_rt->slave_rt_list, struct sdw_slave_runtime, m_rt_node);
+		slave_prop = &s_rt->slave->prop;
+
+		for (i = 1; i < SDW_MAX_LANES; i++) {
+			if (!slave_prop->lane_maps[i])
+				continue;
+
+			required_bandwidth = 0;
+			list_for_each_entry(m_p_rt, &m_rt->port_list, port_node) {
+				required_bandwidth += m_rt->stream->params.rate *
+					hweight32(m_p_rt->ch_mask) *
```

Comment:
@bardliao where are the 'manager ports' ch_mask' defined?

It seems to be that the entire solution is based on a first split of the channels on different ports, but I don't know/understand where that first split is done.

Are we assuming that e.g. the manager code already looks at the bandwidth required in the dai hw_params and relies on multiple ports added with sdw_stream_add_master().

The other problem I have on the manager side is that we have a 1:1 mapping between PDI/port and DAI, so if we need to use multiple ports we would need multiple PDIs per dai.

Or it's possible that for now the solution moves entire streams to different lanes, but that's not solving the 8ch 192kHz problem where multi-lane is required.

Reply:
> @bardliao where are the 'manager ports' ch_mask' defined?
> 
> It seems to be that the entire solution is based on a first split of the channels on different ports, but I don't know/understand where that first split is done.

Not exactly, the solution is not based on splitting channels to different ports. The solution is to check if current lane still has bandwidth for the new port, and use a different lane if there is no enough available bandwidth for the new port. So I don't split the channels to different port in this PR.

> 
> Are we assuming that e.g. the manager code already looks at the bandwidth required in the dai hw_params and relies on multiple ports added with sdw_stream_add_master().

Yes, if I expect that if a single stream needs more than 1 ports, it will be done in the dai hw_params. I am not sure if the manager code need to know and calculate the bandwidth accurately. In fact, I would expect that HDMI will use 2 PDIs/ports unconditionally.

> 
> The other problem I have on the manager side is that we have a 1:1 mapping between PDI/port and DAI, so if we need to use multiple ports we would need multiple PDIs per dai.

Yeah, PDI/port and DAI doesn't have to be 1:1 mapping.

> 
> Or it's possible that for now the solution moves entire streams to different lanes, but that's not solving the 8ch 192kHz problem where multi-lane is required.

Currently, we use a lane per stream. But that is because we use 1 port per stream. We will assign different lanes for different ports in the 8ch 192kHz case if we assign 2 ports for that stream.

Reply:
Humm, I think we're doing this the wrong way, we should first support multiple ports per stream and then add multi-lane support.
The main reason is that your multi-lane parts loops over multiple ports, which isn't going to be tested fully *until* we do have multiple ports.

And the problem is that on paper the DAI-PDI does not need to be 1:1, but that's how it was used so far. I don't know how we would 'reserve' specific PDIs for specific DAIs...

Reply:
> Humm, I think we're doing this the wrong way, we should first support multiple ports per stream and then add multi-lane support. The main reason is that your multi-lane parts loops over multiple ports, which isn't going to be tested fully _until_ we do have multiple ports.

We can test with multi-function codecs like rt722. We will need multi-lane when we run speaker playback, headset playback/capture, and dmic capture simultaneously with 4.8MHz clock rate.

> 
> And the problem is that on paper the DAI-PDI does not need to be 1:1, but that's how it was used so far. I don't know how we would 'reserve' specific PDIs for specific DAIs...

Reply:
Yeah but the code has loops over multiple ports that will only be used with one port. That's why I think we absolutely need in addition to support multiple ports per dai, it's a miss today.